### PR TITLE
refactor(common/web): context-swapping for PredictionContext 🧩

### PR DESCRIPTION
**Pull request by jahorton**
With the direction that the modularization design is heading, things look to be significantly cleaner if `PredictionContext` can be specified and initialized once early on, rather than constantly reinitialized on every change of the context.  That way, a single instance of this class can be passed off as part of a `ContextManager`'s implementation to encapsulate the prediction-related aspects of context, which involve references to objects that `ContextManager` would otherwise not need to access.  

There's way less event-handler churn this way, and it also makes integration with the OSK (see #8466) simpler too.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/input-processor/tests/cases/predictionContext.js [hunk 1]
```diff
@@ -71,22 +71,20 @@ describe("PredictionContext", () => {
...
 
     let mock = new Mock("appl", 4); // "appl|", with '|' as the caret position.
     const initialMock = Mock.from(mock);
-
-    let context = new PredictionContext(langProcessor, kbdProcessor, mock);
-
-    let updateFake = sinon.fake();
-    context.on('update', updateFake);
-    context.sendUpdateEvent(); // Allows external code to request a re-retrieval of current suggestion state.
+    const promise = predictiveContext.setCurrentTarget(mock);
```
Comment (jahorton) [inline common/web/input-processor/tests/cases/predictionContext.js:81]: A bit of happy fallout - I think this feels cleaner and simpler when it comes to unit testing than the prior pattern.

File: common/web/input-processor/src/text/prediction/languageProcessor.ts [hunk 2]
```diff
@@ -66,6 +66,13 @@ interface LanguageProcessorEventMap {
   'statechange': StateChangeHandler,
   'tryaccept': TryUIHandler,
   'tryrevert': () => void,
+
+  /**
+   * Is called synchronously as part of suggestion application when successful.
```
Comment (mcdurdin) [inline common/web/input-processor/src/text/prediction/languageProcessor.ts:71]: When exactly is this called; for example, it'd be helpful to know whether the text buffer in the app has been updated

File: common/web/input-processor/src/text/prediction/predictionContext.ts [hunk 3]
```diff
@@ -37,93 +31,84 @@ export default class PredictionContext extends EventEmitter<PredictionContextEve
...
       throw new Error("Invalid state:  no predictive-text model is currently available.");
     }
+    this.langProcessor = langProcessor;
+    this.kbdProcessor = kbdProcessor;
 
-    if(!target) {
-      throw new Error("No text context is available to use for prediction!");
+    const validSuggestionState: () => boolean = () => {
+      return this.currentTarget && langProcessor.state == 'configured';
     }
```
Comment (mcdurdin) [inline common/web/input-processor/src/text/prediction/predictionContext.ts:82]:
```suggestion
    const validSuggestionState: () => boolean = () => 
      this.currentTarget && langProcessor.state == 'configured';
```

Should also work yeah? No big deal though

**Review by mcdurdin — APPROVED** [review]
LGTM I guess? I don't really understand all the changes happening though.

File: common/web/input-processor/src/text/prediction/languageProcessor.ts (hunk 2 above)
Comment (jahorton) [reply common/web/input-processor/src/text/prediction/languageProcessor.ts:71]: Is this more clear?

```suggestion
   * Is called synchronously once suggestion application is successful and the context has been updated.
```

Comment (mcdurdin) [reply common/web/input-processor/src/text/prediction/languageProcessor.ts:71]: Yep, looks good.

**Pull request merged**